### PR TITLE
website/integrations: Improve discord role sync policies

### PR DESCRIPTION
**Pull request by Keyinator**
## Details

This changes the websites integration page for discord by modifying two policies.

The new policies are easier to understand and take advantage of group attributes instead of hardcoding values (besides the guild id).

---

## Checklist

-   [ ] Local tests pass (`ak test authentik/`)
-   [ ] The code has been formatted (`make lint-fix`)

If an API change has been made

-   [ ] The API schema has been updated (`make gen-build`)

If changes to the frontend have been made

-   [ ] The code has been formatted (`make web`)

If applicable

-   [X] The documentation has been updated
-   [ ] The documentation has been formatted (`make website`)

### CODE REVIEW

**Comment by burritosoftware** [comment]
I tested these last week with @Keyinator over Discord and they work great!

**Review by burritosoftware — APPROVED** [review]

File: website/docs/sources/discord/index.md [hunk 1]
```diff
@@ -158,35 +158,36 @@ Now bind this policy to the chosen enrollment and authentication flows for the D
 ### Syncing Discord roles to authentik groups
 
 :::info
-Ensure that the Discord OAuth source in 'Federation & Social login' has the additional `guilds.members.read` scopes added under the 'Protocol settings'.
+Ensure that the Discord OAuth source in 'Federation & Social login' has the additional `guilds.members.read` scopes added under 'Protocol settings'.
 :::
 
+:::info
+Any authentik role you want to sync with a discord role needs to have the attribute `discord_role_id` with a value of the discord role's id set.
```
Comment (tanberry) [inline website/docs/sources/discord/index.md:165]:
```suggestion
Any authentik role that you want to sync with a Discord role needs to have the attribute `discord_role_id` with a value of the Discord role's ID set.  
```

File: website/docs/sources/discord/index.md [hunk 2]
```diff
@@ -158,35 +158,36 @@ Now bind this policy to the chosen enrollment and authentication flows for the D
...
+Ensure that the Discord OAuth source in 'Federation & Social login' has the additional `guilds.members.read` scopes added under 'Protocol settings'.
 :::
 
+:::info
+Any authentik role you want to sync with a discord role needs to have the attribute `discord_role_id` with a value of the discord role's id set.  
+The settings can be found at `Authentik > Admin Interface > Directory > Groups > YOUR_GROUP > Attributes`  
+Blueprint attribute: ``discord_role_id: "<ROLE ID>"``
+:::
+
+The following two policies will allow you to synchronize roles in a discord guild with roles in authentik.
```
Comment (tanberry) [inline website/docs/sources/discord/index.md:170]:
```suggestion
The following two policies allow you to synchronize roles in a Discord guild with roles in authentik.  
```

File: website/docs/sources/discord/index.md [hunk 3]
```diff
@@ -158,35 +158,36 @@ Now bind this policy to the chosen enrollment and authentication flows for the D
...
 :::
 
+:::info
+Any authentik role you want to sync with a discord role needs to have the attribute `discord_role_id` with a value of the discord role's id set.  
+The settings can be found at `Authentik > Admin Interface > Directory > Groups > YOUR_GROUP > Attributes`  
+Blueprint attribute: ``discord_role_id: "<ROLE ID>"``
+:::
+
+The following two policies will allow you to synchronize roles in a discord guild with roles in authentik.  
+Whenever a user enrolls or signs in to authentik via discord source, these policies will check the user's discord roles and apply the user's authentik roles accordingly.
```
Comment (tanberry) [inline website/docs/sources/discord/index.md:171]:
```suggestion
Whenever a user enrolls or signs in to authentik via a Discord source, these policies will check the user's Discord roles and apply the user's authentik roles accordingly.  
```

File: website/docs/sources/discord/index.md [hunk 4]
```diff
@@ -158,35 +158,36 @@ Now bind this policy to the chosen enrollment and authentication flows for the D
...
 
+:::info
+Any authentik role you want to sync with a discord role needs to have the attribute `discord_role_id` with a value of the discord role's id set.  
+The settings can be found at `Authentik > Admin Interface > Directory > Groups > YOUR_GROUP > Attributes`  
+Blueprint attribute: ``discord_role_id: "<ROLE ID>"``
+:::
+
+The following two policies will allow you to synchronize roles in a discord guild with roles in authentik.  
+Whenever a user enrolls or signs in to authentik via discord source, these policies will check the user's discord roles and apply the user's authentik roles accordingly.  
+All roles with the attribute ``discord_role_id`` defined will be added or removed depending on whether the user is a member of the defined discord role.
```
Comment (tanberry) [inline website/docs/sources/discord/index.md:172]:
```suggestion
All roles with the attribute ``discord_role_id`` defined will be added or removed depending on whether the user is a member of the defined Discord role.
```

**Review by tanberry — CHANGES_REQUESTED** [review]
Thanks for this contribution @Keyinator ! A couple of nits about capitalizing Discord, but yeah as soon as those are made let's merge!

File: website/docs/sources/discord/index.md [hunk 5]
```diff
@@ -158,35 +158,36 @@ Now bind this policy to the chosen enrollment and authentication flows for the D
...
 GUILD_API_URL = "https://discord.com/api/users/@me/guilds/{guild_id}/member"
 
+### CONFIG ###
+guild_id = "<YOUR GUILD ID>"
+##############
+
 # Ensure flow is only run during OAuth logins via Discord
-if context["source"].provider_type != "discord":
-    return True
+if context['source'].provider_type != "discord":
```
Comment (rissson) [inline website/docs/sources/discord/index.md:189]:
```suggestion
if context["source"].provider_type != "discord":
```

File: website/docs/sources/discord/index.md [hunk 6]
```diff
@@ -195,28 +196,36 @@ if not connection:
...
+  GUILD_API_URL.format(guild_id=guild_id),
+  headers={
+    "Authorization": "Bearer " + access_token
+  },
 ).json()
 
 # Ensure user is a member of the guild
 if "code" in guild_member_info:
-    if guild_member_info["code"] == 10004:
+    if guild_member_info['code'] == 10004:
```
Comment (rissson) [inline website/docs/sources/discord/index.md:207]:
```suggestion
    if guild_member_info["code"] == 10004:
```

File: website/docs/sources/discord/index.md [hunk 7]
```diff
@@ -195,28 +196,36 @@ if not connection:
 access_token = connection.access_token
 
 guild_member_info = requests.get(
-    GUILD_API_URL.format(guild_id=GUILD_ID),
-    headers={"Authorization": "Bearer " + access_token},
+  GUILD_API_URL.format(guild_id=guild_id),
+  headers={
+    "Authorization": "Bearer " + access_token
```
Comment (rissson) [inline website/docs/sources/discord/index.md:201]:
```suggestion
    "Authorization": f"Bearer {access_token}",
```

File: website/docs/sources/discord/index.md [hunk 8]
```diff
@@ -255,35 +255,44 @@ if not connection:
...
 # Ensure user is a member of the guild
 if "code" in guild_member_info:
-    if guild_member_info["code"] == 10004:
+    if guild_member_info['code'] == 10004:
         ak_message("User is not a member of the guild")
     else:
-        ak_create_event(
-            "discord_error", source=context["source"], code=guild_member_info["code"]
-        )
+        ak_create_event("discord_error", source=context['source'], code=guild_member_info['code'])
```
Comment (rissson) [inline website/docs/sources/discord/index.md:269]:
```suggestion
        ak_create_event("discord_error", source=context["source"], code=guild_member_info["code"])
```

File: website/docs/sources/discord/index.md [hunk 9]
```diff
@@ -195,28 +196,36 @@ if not connection:
...
```
Comment (rissson) [inline website/docs/sources/discord/index.md:210]:
```suggestion
        ak_create_event("discord_error", source=context["source"], code=guild_member_info["code"])
```

File: website/docs/sources/discord/index.md [hunk 10]
```diff
@@ -255,35 +255,44 @@ if not connection:
...
+# Filter matching roles based on guild_member_info['roles']
+matching_roles = discord_groups.filter(attributes__discord_role_id__in=guild_member_info['roles'])
+
+# Combine user_groups and matching_roles
+combined_groups = user_groups.union(matching_roles)
 
 # Update user's groups
-request.user.ak_groups.set(new_groups)
+request.user.ak_groups.set(combined_groups)
 request.user.save()
```
Comment (rissson) [inline website/docs/sources/discord/index.md:287]: Saving the user here should not be needed

Comment (rissson) [reply website/docs/sources/discord/index.md:287]:
```suggestion
```

File: website/docs/sources/discord/index.md [hunk 11]
```diff
@@ -195,28 +196,36 @@ if not connection:
...
-# Add all mapped roles the user has in the guild
-groups_to_add = []
-for role_id in MAPPED_ROLES:
-    if role_id in guild_member_info["roles"]:
-        groups_to_add.append(MAPPED_ROLES[role_id])
+# Get all discord_groups
+discord_groups = Group.objects.filter(attributes__discord_role_id__isnull=False)
+
+# Filter matching roles based on guild_member_info['roles']
+matching_roles = discord_groups.filter(attributes__discord_role_id__in=guild_member_info['roles'])
```
Comment (rissson) [inline website/docs/sources/discord/index.md:218]:
```suggestion
matching_roles = discord_groups.filter(attributes__discord_role_id__in=guild_member_info["roles"])
```

File: website/docs/sources/discord/index.md [hunk 12]
```diff
@@ -195,28 +196,36 @@ if not connection:
...
+
+# Set matchin_roles in flow context
+request.context["flow_plan"].context["groups"] = matching_roles
+
+# Create event with roles added
+ak_create_event(
+    "discord_role_sync",
+    discord_roles_added=', '.join(str(group) for group in matching_roles),
```
Comment (rissson) [inline website/docs/sources/discord/index.md:226]:
```suggestion
    discord_roles_added=", ".join(str(group) for group in matching_roles),
```

File: website/docs/sources/discord/index.md [hunk 13]
```diff
@@ -255,35 +255,44 @@ if not connection:
...
 
-# Get the user's current roles and remove all roles we want to remap
-new_groups = [
-    role for role in request.user.ak_groups.all() if role not in MAPPED_ROLES.values()
-]
+# Get all discord_groups
+discord_groups = Group.objects.filter(attributes__discord_role_id__isnull=False)
+
+# Get all user groups except discord_groups
+user_groups = request.user.ak_groups.exclude(pk__in=discord_groups.values_list('pk', flat=True))
```
Comment (rissson) [inline website/docs/sources/discord/index.md:277]:
```suggestion
user_groups = request.user.ak_groups.exclude(pk__in=discord_groups.values_list("pk", flat=True))
```

File: website/docs/sources/discord/index.md [hunk 14]
```diff
@@ -255,35 +255,44 @@ if not connection:
...
 
-# Add back mapped roles which the user has in the guild
-for role_id in MAPPED_ROLES:
-    if role_id in guild_member_info["roles"]:
-        new_groups.append(MAPPED_ROLES[role_id])
+# Filter matching roles based on guild_member_info['roles']
+matching_roles = discord_groups.filter(attributes__discord_role_id__in=guild_member_info['roles'])
```
Comment (rissson) [inline website/docs/sources/discord/index.md:280]:
```suggestion
matching_roles = discord_groups.filter(attributes__discord_role_id__in=guild_member_info["roles"])
```

File: website/docs/sources/discord/index.md [hunk 15]
```diff
@@ -255,35 +255,44 @@ if not connection:
...
 # Update user's groups
-request.user.ak_groups.set(new_groups)
+request.user.ak_groups.set(combined_groups)
 request.user.save()
 
+# Create event with roles changed
+ak_create_event(
+    "discord_role_sync",
+    discord_roles_before=', '.join(str(group) for group in request.user.ak_groups.filter(pk__in=discord_groups.values_list('pk', flat=True))),
+    discord_roles_after=', '.join(str(group) for group in matching_roles),
```
Comment (rissson) [inline website/docs/sources/discord/index.md:293]:
```suggestion
    discord_roles_before=", ".join(str(group) for group in request.user.ak_groups.filter(pk__in=discord_groups.values_list("pk", flat=True))),
    discord_roles_after=", ".join(str(group) for group in matching_roles),
```

Comment (rissson) [reply website/docs/sources/discord/index.md:293]: Those queries should already be available from above. Why not reuse them instead of doing additional database queries?

**Review by rissson — CHANGES_REQUESTED** [review]

File: website/docs/sources/discord/index.md (hunk 15 above)
Comment (Keyinator) [reply website/docs/sources/discord/index.md]: > Those queries should already be available from above. Why not reuse them instead of doing additional database queries?

Do you mean ``request.user_ak_groups`` or ``request.user.ak_groups.filter(pk__in=discord_groups.values_list('pk', flat=True))``?

If you are saying that ``request.user.ak_groups`` is calculated on every access, we can definitely do that.
If you are referring to the value of ``discord_roles_before`` then the following should explain why we can't reuse above code:

In the code preceeding the event we filter out roles which are __not__ in `discord_groups`.
However in this event, we want to display `discord_groups` the user was in.

Basically we have these relations:
`User-Roles` - `discord_groups` (preceeding code)
`User-Roles` ∩ `discord_groups` (within event)

If I am not mistaken, one cannot easily be reused for the other

Please let me know which you mean so I can adjust the code.

Comment (rissson) [reply website/docs/sources/discord/index.md]: So actually, the second one is fine, but the first one, `discord_roles_before` is I think wrong. As written, the query is based on `request.user.ak_groups` _after_ it's already been modified. So if any roles were removed, they wouldn't show up in the event.

Comment (Keyinator) [reply website/docs/sources/discord/index.md]: > So actually, the second one is fine, but the first one, `discord_roles_before` is I think wrong. As written, the query is based on `request.user.ak_groups` _after_ it's already been modified. So if any roles were removed, they wouldn't show up in the event.

I checked and it seems to be working fine.
I suppose the request-object is just the in-time data when the policy is executed.

File: website/docs/sources/discord/index.md (hunk 10 above)
Comment (Keyinator) [reply website/docs/sources/discord/index.md:287]: You are right.

File: website/docs/sources/discord/index.md (hunk 15 above)
Comment (rissson) [reply website/docs/sources/discord/index.md]: No, it will actually execute the `request.user.ak_groups.filter(pk__in=discord_groups.values_list("pk", flat=True))` query after you've modified the data. If we really want to log proper data, that query should be done before changing ak_groups, saved as a variable and used when needed.

File: website/docs/sources/discord/index.md [hunk 16]
```diff
@@ -202,6 +202,11 @@ guild_member_info = requests.get(
   },
 ).json()
 
+# Ensure we are not being ratelimited
+if guild_member_info.get("message") == "You are being rate limited.":
```
Comment (rissson) [inline website/docs/sources/discord/index.md:206]: This is quite weak as a check. It would be better to:

```python3
guild_member_request = requests.get(
  GUILD_API_URL.format(guild_id=guild_id),
  headers={
    "Authorization": f"Bearer {access_token}",
  },
)
guild_member_info = guild_member_request.json()

# Ensure we are not being ratelimited
if guild_member_request.status_code == 429:
	pass # do things
```

And the same below.

Cf. https://discord.com/developers/docs/topics/rate-limits

File: website/docs/sources/discord/index.md (hunk 15 above)
Comment (Keyinator) [reply website/docs/sources/discord/index.md]: > No, it will actually execute the `request.user.ak_groups.filter(pk__in=discord_groups.values_list("pk", flat=True))` query after you've modified the data. If we really want to log proper data, that query should be done before changing ak_groups, saved as a variable and used when needed.

I tested it on 2024.4.3 and ``discord_roles_before`` was showing correctly. But we can change it if you prefer.

File: website/docs/sources/discord/index.md (hunk 16 above)
Comment (Keyinator) [reply website/docs/sources/discord/index.md:206]: I am not sure if guild_member_request has this info. When I sent guild_member_info as an event it only had the following properties:
```
"guild_member_info": {
        "global": false,
        "message": "You are being rate limited.",
        "retry_after": 35.683
    },
```

File: website/docs/sources/discord/index.md (hunk 15 above)
Comment (rissson) [reply website/docs/sources/discord/index.md]: Yeah, I'd rather we change it, so we're 100% sure it works

File: website/docs/sources/discord/index.md (hunk 16 above)
Comment (rissson) [reply website/docs/sources/discord/index.md:206]: So it goes like this:
```python
guild_member_request = requests.get(...)
# This is a requests.Response object, which has, among a lot of other things, a status_code attribute and a json() method.
guild_member_info = guild_member_request.json()
# This takes the requests.Response.body and passes it to json.loads and gives it back.
```
So from `guild_member_request`, you can get the HTTP status code to check you're not being rate limited, and then from `guild_member_info` you can get either the "roles" key (and others) if you're not being rate limited, or the `retry_after` key if you are.

**Comment by tanberry** [comment]
HI @Keyinator can you run `make website` or `npm run prettier` again and repush? That build check is failing. We'll get this merged yet!! :-)

**Comment by Keyinator** [comment]
> HI @Keyinator can you run `make website` or `npm run prettier` again and repush? That build check is failing. We'll get this merged yet!! :-)

There were indeed fails in my file (fixed now).  
However there is also lint-issues in `website/integrations/services/gitlab/index.md`. How do we proceed?

**Comment by rissson** [comment]
> There were indeed fails in my file (fixed now). However, there is also lint-issues in `website/integrations/services/gitlab/index.md`. How do we proceed?

You can `git fetch && git merge origin/main`, which should fix the issue in this particular branch. I suppose it was created when we had those fails on `main`.

**Comment by rissson** [comment]
I'm sorry we broke the main CI again (we're having GitHub actions issues with this currently). #10524 should fix it. Once that's merged, could you merge main into this branch again?

**Comment by Keyinator** [comment]
> I'm sorry we broke the main CI again (we're having GitHub actions issues with this currently). #10524 should fix it. Once that's merged, could you merge main into this branch again?

Sure thing :)

**Review by rissson — APPROVED** [review]

File: website/docs/sources/discord/index.md [hunk 17]
```diff
@@ -158,65 +158,80 @@ Now bind this policy to the chosen enrollment and authentication flows for the D
...
 GUILD_API_URL = "https://discord.com/api/users/@me/guilds/{guild_id}/member"
 
+### CONFIG ###
+guild_id = "<YOUR GUILD ID>"
+##############
+
 # Ensure flow is only run during OAuth logins via Discord
 if context["source"].provider_type != "discord":
-    return True
+  return True
```
Comment (rissson) [inline website/docs/sources/discord/index.md:190]:
```suggestion
    return True
```

File: website/docs/sources/discord/index.md [hunk 18]
```diff
@@ -158,65 +158,80 @@ Now bind this policy to the chosen enrollment and authentication flows for the D
...
 
-guild_member_info = requests.get(
-    GUILD_API_URL.format(guild_id=GUILD_ID),
-    headers={"Authorization": "Bearer " + access_token},
-).json()
+guild_member_request = requests.get(
+  GUILD_API_URL.format(guild_id=guild_id),
+  headers={
+    "Authorization": f"Bearer {access_token}",
+  },
```
Comment (rissson) [inline website/docs/sources/discord/index.md:202]:
```suggestion
    GUILD_API_URL.format(guild_id=guild_id),
    headers={
        "Authorization": f"Bearer {access_token}",
    },
```

File: website/docs/sources/discord/index.md [hunk 19]
```diff
@@ -158,65 +158,80 @@ Now bind this policy to the chosen enrollment and authentication flows for the D
...
+)
+guild_member_info = guild_member_request.json()
+
+# Ensure we are not being ratelimited
+if guild_member_request.status_code == 429:
+  ak_message(f"Discord is throttling this connection. Retry in {int(guild_member_info['retry_after'])}s")
+  return False
```
Comment (rissson) [inline website/docs/sources/discord/index.md:209]:
```suggestion
    ak_message(f"Discord is throttling this connection. Retry in {int(guild_member_info['retry_after'])}s")
    return False
```

File: website/docs/sources/discord/index.md [hunk 20]
```diff
@@ -254,35 +260,50 @@ if not connection:
...
-guild_member_info = requests.get(
-    GUILD_API_URL.format(guild_id=GUILD_ID),
-    headers={"Authorization": "Bearer " + access_token},
-).json()
+guild_member_request = requests.get(
+  GUILD_API_URL.format(guild_id=guild_id),
+  headers={
+    "Authorization": f"Bearer {access_token}"
+  },
+)
```
Comment (rissson) [inline website/docs/sources/discord/index.md:268]:
```suggestion
    GUILD_API_URL.format(guild_id=guild_id),
    headers={
        "Authorization": f"Bearer {access_token}"
    },
)
```

File: website/docs/sources/discord/index.md [hunk 21]
```diff
@@ -254,35 +260,50 @@ if not connection:
...
+guild_member_info = guild_member_request.json()
+
+# Ensure we are not being ratelimited
+if guild_member_request.status_code == 429:
+  ak_message(f"Discord is throttling this connection. Retry in {int(guild_member_info['retry_after'])}s")
+  return False
```
Comment (rissson) [inline website/docs/sources/discord/index.md:274]:
```suggestion
    ak_message(f"Discord is throttling this connection. Retry in {int(guild_member_info['retry_after'])}s")
    return False
```

**Review by rissson — APPROVED** [review]

**Review by rissson — APPROVED** [review]

**Review by tanberry — APPROVED** [review]
